### PR TITLE
Admin borg panel

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
It's a little bit WIP but everything works apart from the player panel link, I'll have to check that later

:cl: Naksu
admin: Admins can now access a new control panel for borgs via a new admin verb
/:cl:

image

### CODE REVIEW

**Comment by vuonojenmustaturska** [comment]
Gonna have to refactor borg upgrades a bit

**Comment by Bluespace** [comment]
This is a really nice and easy to use interface. good on you.

**Comment by Bombpants** [comment]
Any chance of adding upgrades to it. Things like higher capacity battery or radio keys.

**Comment by vuonojenmustaturska** [comment]
@Bombpants yes i already did both, the screenshot is out of date

**Comment by vuonojenmustaturska** [comment]
updated op with new screenshot

**Comment by vuonojenmustaturska** [comment]
realized the usr->user work isn't done and a button is missing.

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 1]
```diff
@@ -14,14 +14,20 @@
 	// if module is reset
 	var/one_use = FALSE
 
-/obj/item/borg/upgrade/proc/action(mob/living/silicon/robot/R)
+/obj/item/borg/upgrade/proc/action(mob/living/silicon/robot/R, user = usr)
```
Comment (Cyberboss) [inline code/game/objects/items/robot/robot_upgrades.dm]: mob/user

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 2]
```diff
@@ -14,14 +14,20 @@
...
+		return FALSE
 	if(module_type && !istype(R.module, module_type))
 		to_chat(R, "Upgrade mounting error!  No suitable hardpoint detected!")
-		to_chat(usr, "There's no mounting point for the module!")
-		return 1
+		to_chat(user, "There's no mounting point for the module!")
+		return FALSE
+	return TRUE
+
+/obj/item/borg/upgrade/proc/deactivate(mob/living/silicon/robot/R, user = usr)
```
Comment (Cyberboss) [inline code/game/objects/items/robot/robot_upgrades.dm]: ditto

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 3]
```diff
@@ -34,55 +40,51 @@
...
-	R.custom_name = heldname
-	R.updatename()
-	if(oldname == R.real_name)
-		R.notify_ai(RENAME, oldname, R.real_name)
-
-	return 1
-
+	. = ..()
+	if(.)
+		var/oldname = R.real_name
```
Comment (Cyberboss) [inline code/game/objects/items/robot/robot_upgrades.dm]: early return

Comment (vuonojenmustaturska) [reply code/game/objects/items/robot/robot_upgrades.dm]: the code is literally equivalent to ```
. = ..()
if (!.)
  return
<code>
```

**Review by imtakingabreakdontatme — DISMISSED** [review]

**Comment by KorPhaeron** [comment]
Looks good to me but I'm gonna give cyberboss a day or two to respond to what you told him about the reviews before hitting the button

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/reagents/reagent_containers/borghydro.dm [hunk 4]
```diff
@@ -67,6 +67,18 @@ Borg Hypospray
 
 	modes[reagent] = modes.len + 1
 
+/obj/item/reagent_containers/borghypo/proc/del_reagent(reagent)
+	reagent_ids -= reagent
+	var/datum/reagents/RG
+	for(var/i in 1 to reagent_ids.len)
+		RG = reagent_list[i]
```
Comment (Cyberboss) [inline code/modules/reagents/reagent_containers/borghydro.dm]: This will always delete something even if it doesn't have reagent. Intended?

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/borgpanel.dm [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		if ("slavetoai")
+			var/mob/living/silicon/ai/newai = locate(params["slavetoai"]) in GLOB.ai_list
+			if (newai && newai != borg.connected_ai)
+				borg.notify_ai(DISCONNECT)
+				if(borg.shell)
+					borg.undeploy()
+				borg.connected_ai = newai
+				borg.notify_ai(TRUE)
+				message_admins("[key_name_admin(user)] slaved [key_name_admin(borg)][ADMIN_FLW(borg)] to the AI [key_name_admin(newai)][ADMIN_FLW(newai)].")
+				log_admin("[key_name_admin(user)] slaved [key_name_admin(borg)] to the AI [key_name_admin(newai)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: key_name

File: code/modules/admin/verbs/borgpanel.dm [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+					borg.radio.keyslot.independent = TRUE
+				message_admins("[key_name_admin(user)] added the [channel] radio channel to [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] added the [channel] radio channel to [key_name_admin(borg)].")
+			borg.radio.recalculateChannels()
+		if ("setmodule")
+			var/newmodulepath = text2path(params["module"])
+			if (ispath(newmodulepath))
+				borg.module.transform_to(newmodulepath)
+				message_admins("[key_name_admin(user)] changed the module of [key_name_admin(borg)][ADMIN_FLW(borg)] to [newmodulepath].")
+				log_admin("[key_name_admin(user)] changed the module of [key_name_admin(borg)] to [newmodulepath].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: key_name

File: code/modules/admin/verbs/borgpanel.dm [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+			else	// We're adding a channel
+				if (!borg.radio.keyslot) // Assert that an encryption key exists
+					borg.radio.keyslot = new (borg.radio)
+				borg.radio.keyslot.channels[channel] = 1
+				if (channel == "Syndicate")
+					borg.radio.keyslot.syndie = TRUE
+				else if (channel == "CentCom")
+					borg.radio.keyslot.independent = TRUE
+				message_admins("[key_name_admin(user)] added the [channel] radio channel to [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] added the [channel] radio channel to [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: key_name

File: code/modules/admin/verbs/borgpanel.dm [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+					else if (channel == "CentCom")
+						borg.radio.independent = FALSE
+				else
+					borg.radio.keyslot.channels -= channel
+					if (channel == "Syndicate")
+						borg.radio.keyslot.syndie = FALSE
+					else if (channel == "CentCom")
+						borg.radio.keyslot.independent = FALSE
+				message_admins("[key_name_admin(user)] removed the [channel] radio channel from [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] removed the [channel] radio channel from [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+				borg.upgrades -= installedupgrade
+				message_admins("[key_name_admin(user)] removed the [installedupgrade] upgrade from [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] removed the [installedupgrade] upgrade from [key_name_admin(borg)].")
+				qdel(installedupgrade)
+			else
+				var/obj/item/borg/upgrade/upgrade = new upgradepath(borg)
+				upgrade.action(borg, user)
+				borg.upgrades += upgrade
+				message_admins("[key_name_admin(user)] added the [upgrade] borg upgrade to [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] added the [upgrade] borg upgrade to [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+			log_admin("[key_name_admin(user)] renamed [key_name_admin(borg)] to [new_name].")
+			borg.fully_replace_character_name(borg.real_name,new_name)
+		if ("toggle_upgrade")
+			var/upgradepath = text2path(params["upgrade"])
+			var/obj/item/borg/upgrade/installedupgrade = locate(upgradepath) in borg
+			if (installedupgrade)
+				installedupgrade.deactivate(borg, user)
+				borg.upgrades -= installedupgrade
+				message_admins("[key_name_admin(user)] removed the [installedupgrade] upgrade from [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] removed the [installedupgrade] upgrade from [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+				log_admin("[key_name_admin(user)] enabled scrambled codes on [key_name_admin(borg)].")
+			else
+				message_admins("[key_name_admin(user)] disabled scrambled codes on [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] disabled scrambled codes on [key_name_admin(borg)].")
+		if ("rename")
+			var/new_name = stripped_input(user,"What would you like to name this cyborg?","Input a name",borg.real_name,MAX_NAME_LEN)
+			if(!new_name)
+				return
+			message_admins("[key_name_admin(user)] renamed [key_name_admin(borg)][ADMIN_FLW(borg)] to [new_name].")
+			log_admin("[key_name_admin(user)] renamed [key_name_admin(borg)] to [new_name].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+				message_admins("[key_name_admin(user)] locked down [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] locked down [key_name_admin(borg)].")
+			else
+				message_admins("[key_name_admin(user)] released [key_name_admin(borg)][ADMIN_FLW(borg)] from lockdown.")
+				log_admin("[key_name_admin(user)] released [key_name_admin(borg)] from lockdown.")
+		if ("toggle_scrambledcodes")
+			borg.scrambledcodes = !borg.scrambledcodes
+			if (borg.scrambledcodes)
+				message_admins("[key_name_admin(user)] enabled scrambled codes on [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] enabled scrambled codes on [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+				message_admins("[key_name_admin(user)] disabled lawsync on [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] disabled lawsync on [key_name_admin(borg)].")
+		if ("toggle_lockdown")
+			borg.SetLockdown(!borg.lockcharge)
+			if (borg.lockcharge)
+				message_admins("[key_name_admin(user)] locked down [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] locked down [key_name_admin(borg)].")
+			else
+				message_admins("[key_name_admin(user)] released [key_name_admin(borg)][ADMIN_FLW(borg)] from lockdown.")
+				log_admin("[key_name_admin(user)] released [key_name_admin(borg)] from lockdown.")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+				message_admins("[key_name_admin(user)] enabled lawsync on [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] enabled lawsync on [key_name_admin(borg)].")
+			else
+				message_admins("[key_name_admin(user)] disabled lawsync on [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] disabled lawsync on [key_name_admin(borg)].")
+		if ("toggle_lockdown")
+			borg.SetLockdown(!borg.lockcharge)
+			if (borg.lockcharge)
+				message_admins("[key_name_admin(user)] locked down [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] locked down [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+				message_admins("[key_name_admin(user)] emagged [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] emagged [key_name_admin(borg)].")
+			else
+				message_admins("[key_name_admin(user)] un-emagged [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] un-emagged [key_name_admin(borg)].")
+		if ("toggle_lawupdate")
+			borg.lawupdate = !borg.lawupdate
+			if (borg.lawupdate)
+				message_admins("[key_name_admin(user)] enabled lawsync on [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] enabled lawsync on [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+				message_admins("[key_name_admin(user)] changed the cell of [key_name_admin(borg)][ADMIN_FLW(borg)] to [new_cell].")
+				log_admin("[key_name_admin(user)] changed the cell of [key_name_admin(borg)] to [new_cell].")
+		if ("toggle_emagged")
+			borg.SetEmagged(!borg.emagged)
+			if (borg.emagged)
+				message_admins("[key_name_admin(user)] emagged [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] emagged [key_name_admin(borg)].")
+			else
+				message_admins("[key_name_admin(user)] un-emagged [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] un-emagged [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+				borg.cell = new_cell
+				borg.cell.charge = borg.cell.maxcharge
+				borg.diag_hud_set_borgcell()
+				message_admins("[key_name_admin(user)] changed the cell of [key_name_admin(borg)][ADMIN_FLW(borg)] to [new_cell].")
+				log_admin("[key_name_admin(user)] changed the cell of [key_name_admin(borg)] to [new_cell].")
+		if ("toggle_emagged")
+			borg.SetEmagged(!borg.emagged)
+			if (borg.emagged)
+				message_admins("[key_name_admin(user)] emagged [key_name_admin(borg)][ADMIN_FLW(borg)].")
+				log_admin("[key_name_admin(user)] emagged [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+				chosen = text2path(chosen)
+			if (chosen)
+				if (borg.cell)
+					QDEL_NULL(borg.cell)
+				var/new_cell = new chosen(borg)
+				borg.cell = new_cell
+				borg.cell.charge = borg.cell.maxcharge
+				borg.diag_hud_set_borgcell()
+				message_admins("[key_name_admin(user)] changed the cell of [key_name_admin(borg)][ADMIN_FLW(borg)] to [new_cell].")
+				log_admin("[key_name_admin(user)] changed the cell of [key_name_admin(borg)] to [new_cell].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+		if ("set_charge")
+			var/newcharge = input("New charge (0-[borg.cell.maxcharge]):", borg.name, borg.cell.charge) as num|null
+			if (newcharge)
+				borg.cell.charge = CLAMP(newcharge, 0, borg.cell.maxcharge)
+				message_admins("[key_name_admin(user)] set the charge of [key_name_admin(borg)][ADMIN_FLW(borg)] to [borg.cell.charge].")
+				log_admin("[key_name_admin(user)] set the charge of [key_name_admin(borg)] to [borg.cell.charge].")
+		if ("remove_cell")
+			QDEL_NULL(borg.cell)
+			message_admins("[key_name_admin(user)] deleted the cell of [key_name_admin(borg)][ADMIN_FLW(borg)].")
+			log_admin("[key_name_admin(user)] deleted the cell of [key_name_admin(borg)].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+/datum/borgpanel/ui_act(action, params)
+	if(..())
+		return
+	switch (action)
+		if ("set_charge")
+			var/newcharge = input("New charge (0-[borg.cell.maxcharge]):", borg.name, borg.cell.charge) as num|null
+			if (newcharge)
+				borg.cell.charge = CLAMP(newcharge, 0, borg.cell.maxcharge)
+				message_admins("[key_name_admin(user)] set the charge of [key_name_admin(borg)][ADMIN_FLW(borg)] to [borg.cell.charge].")
+				log_admin("[key_name_admin(user)] set the charge of [key_name_admin(borg)] to [borg.cell.charge].")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: ^

File: code/modules/admin/verbs/borgpanel.dm [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+		if (k == "Common")
+			continue
+		.["channels"] += list(list("name" = k, "installed" = (k in borg.radio.channels)))
+	.["cell"] = borg.cell ? list("missing" = FALSE, "maxcharge" = borg.cell.maxcharge, "charge" = borg.cell.charge) : list("missing" = TRUE, "maxcharge" = 1, "charge" = 0)
+	.["modules"] = list()
+	for(var/moduletype in typesof(/obj/item/robot_module))
+		var/obj/item/robot_module/module = moduletype
+		.["modules"] += list(list(
+			"name" = initial(module.name),
+			"type" = "[module]"
```
Comment (Cyberboss) [inline code/modules/admin/verbs/borgpanel.dm]: stringification not necessary but w/e

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**